feat(hardware): Load pipette model from pipette serial

sfoster1 opened the pull request:
On the OT-2, we store the literal serial number of a pipette on its EEPROM. When we detect whether it's connected, we read that serial number and parse it out in python.

We could do that here, but we have access to a lot more smarts on the pipette now, so instead let's parse the serial at flash time and store a binary-packed value for the name and model. The datecode part is still stored literally. 

Locally, we can parse it pretty easily, and use it as the model for the rest of the system (which we more or less already do - not many changes on the read side).

The write side is very different. Let's add a script for writing pipette serials but actually make it part of the system this time.

## Testing
- Write and read some serial numbers! You'll need https://github.com/Opentrons/ot3-firmware/pull/437 on the pipette

## TODO
- [x] This needs proper integration tests

Closes RET-1168
Closes RET-1167

sfoster1 committed:
refactor(hardware): flash pipette serials

Adds a new module for dealing with pipette serial numbers and a tool for
flashing them and confirming they're correct.

Closes RET-1168

sfoster1 committed:
add the used names to the enum

sfoster1 committed:
fix some size and byte order issues

Laura-Danielle reviewed (approved):
Nice. LGTM once failing tests are fixed. Also, did you do a sanity check on the integration tests to make sure behavior isn't changing?

sfoster1 commented:
no, that's the todo

sfoster1 committed:
fix tests

sfoster1 committed:
refactor(hardware): cap sensor integration fix

The capacitive sensor on hardware has an offset handler that will
gradually grow if necessary. The simulator harness in firmware would
pretend the sensor was returning a value large enough to make the offset
change, so over time the sensor value returned on can would grow. Making
the value returned by the harness smaller fixes this issue, but means we
have to change the expected values locally.

sfoster1 committed:
refactor(hardware): sensor integration test fixes

The other sensor integration tests were commonizing a little too much
behavior and needed some unique work for each sensor. Now they pass.

sfoster1 committed:
refactor(hardware): fix serial length and integration

sfoster1 committed:
refactor(api): ot3: split models and test

caila-marashaj committed:
add handling for unknown tool names and models

Laura-Danielle commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -380,7 +380,7 @@ async def get_attached_instruments(
             name: FirmwarePipetteName, model: int
         ) -> "PipetteModel":
-            return cast("PipetteModel", name.name + "_v3." + str(model))
+            return cast("PipetteModel", f"{name.name}_v{model//10}.{model%10}")
```
How does this fix the original problem if the pipette is not flashed? It will still give the incorrect model lookup info if I'm reading this correctly.

sfoster1 replied:
that's not what solves it. the concept here is that we previously may have wanted to sort of suppress errors coming from unflashed pipettes, but now we can actually flash them, so instead let's emit a structured error. So, it's the code that catches `KeyErrors` from pipette config and turns them into `UnknownPipetteModel` exceptions that solves the issue.

sfoster1 replied:
Oh the answer to this was: we forgot to commit half the work. Here's the other half that actually does what I said above @Laura-Danielle

Laura-Danielle reviewed (changes requested):
Unless I am reading this wrong, I don't think the original problem of non-flashed pipettes is fully solved.

sfoster1 committed:
Handle unknown pipette names and models

The hardware layer is responsible for making sure that values are inside
correct python types. It makes sure that even if a random int gets
returned for the pipette name, it gets mapped to unknown; and that a
model int gets mapped to a dotted-version string.

Once those exist, the hardware control needs to handle values that are
in the right type, but are semantically invalid. Here, the solution is
to catch the underlying python errors of e.g. "i dont know what this
model means so here's a keyerror" into a semantically relevant exceptions.

Laura-Danielle commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -363,6 +365,59 @@ async def gripper_home_jaw(self) -> None:
+        attached: ohc_tool_types.GripperInformation,
+    ) -> AttachedGripper:
+        model = gripper_config.info_num_to_model(attached.model)
+        serial = attached.serial
```
We'll need to handle the same problem for the gripper as they will start unflashed, otherwise it will hit a KeyError.

Laura-Danielle commented on hardware/opentrons_hardware/pipettes/serials.py:
```diff
@@ -68,3 +68,8 @@ def serial_val_from_parts(name: PipetteName, model: int, serialval: bytes) -> by
+
+def model_versionstring_from_int(model: int) -> str:
+    """Format the encoded model from an int into a dotted version string."""
+    return f"{model//10}.{model%10}"
```
So `6553.5` for unflashed pipettes?

Laura-Danielle reviewed (approved):
Looks good, but we should handle unflashed grippers also while we're here.
